File: IdeaProjects/TugasPraktikum1_BookManager2/src/test/java/BookManagerTest.java
//
// Source code recreated from a .class file by IntelliJ IDEA
// (powered by FernFlower decompiler)
//

import java.util.List;
import org.example.Book;
import org.example.BookManager;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

public class BookManagerTest {
    private BookManager bookManager;

    @BeforeEach
    void setUp() {
        this.bookManager = new BookManager();
    }

    @Test
    @DisplayName("Test menambahkan buku")
    void testAddBook() {
        Book buku = new Book("Pemrograman", "Andi", 2020);
        this.bookManager.addBook(buku);
        Assertions.assertEquals(1, this.bookManager.getBookCount());
    }

    @Test
    @DisplayName("Test menghapus buku yang ada")
    void testRemoveExistingBook() {
        Book buku = new Book("Basis Data", "ErLangga", 2021);
        this.bookManager.addBook(buku);
        boolean removed = this.bookManager.removeBook("Basis Data");
        Assertions.assertTrue(removed);
        Assertions.assertEquals(0, this.bookManager.getBookCount());
    }

    @Test
    @DisplayName("Test menghapus buku yang tidak ada")
    void testRemoveNonExistingBook() {
        boolean removed = this.bookManager.removeBook("Tidak Ada");
        Assertions.assertFalse(removed);
    }

    @Test
    @DisplayName("Test mencari buku berdasarkan author")
    void testFindBooksByAuthor() {
        Book buku = new Book("Java", "Dewo", 2025);
        this.bookManager.addBook(buku);
        List<Book> result = this.bookManager.findBooksByAuthor("Dewo");
        Assertions.assertEquals(1, result.size());
        Assertions.assertEquals("Java", ((Book)result.get(0)).getJudul_Buku());
    }

    @Test
    @DisplayName("Test mendapatkan semua buku")
    void testGetAllBooks() {
        Book buku1 = new Book("Java", "Dewo", 2025);
        Book buku2 = new Book("Python", "Andi", 2024);
        this.bookManager.addBook(buku1);
        this.bookManager.addBook(buku2);
        List<Book> allBooks = this.bookManager.getAllBooks();
        Assertions.assertEquals(2, allBooks.size());
    }
}

File: bookManager.test.js
const Book = require("./book");
const BookManager = require("./BookManager");

describe("BookManager", () => {
  let bookManager;

  beforeEach(() => {
    bookManager = new BookManager();
  });

  test("Test menambahkan buku", () => {
    const book = new Book("Test Book", "Test Author", 2023);
    bookManager.addBook(book);
    expect(bookManager.getBookCount()).toBe(1);
  });

  test("Test menghapus buku yang ada", () => {
    const book = new Book("To Remove", "Author", 2023);
    bookManager.addBook(book);

    const removed = bookManager.removeBook("To Remove");
    expect(removed).toBe(true);
    expect(bookManager.getBookCount()).toBe(0);
  });

  // Test menghapus buku yang tidak ada
  test("Test menghapus buku yang tidak ada", () => {
    const removed = bookManager.removeBook("Not Exist Book");
    expect(removed).toBe(false);
    expect(bookManager.getBookCount()).toBe(0);
  });

  // Test mencari buku berdasarkan penulis
  test("Test mencari buku berdasarkan author", () => {
    const book1 = new Book("Book A", "Author1", 2020);
    const book2 = new Book("Book B", "Author2", 2021);
    const book3 = new Book("Book C", "Author1", 2022);

    bookManager.addBook(book1);
    bookManager.addBook(book2);
    bookManager.addBook(book3);

    const booksByAuthor1 = bookManager.findBooksByAuthor("Author1");

    expect(booksByAuthor1.length).toBe(2);
    expect(booksByAuthor1).toEqual(expect.arrayContaining([book1, book3]));
  });

  // Test mendapatkan semua buku
  test("Test mendapatkan semua buku", () => {
    const book1 = new Book("Book A", "Author1", 2020);
    const book2 = new Book("Book B", "Author2", 2021);

    bookManager.addBook(book1);
    bookManager.addBook(book2);

    const allBooks = bookManager.getAllBooks();

    expect(allBooks.length).toBe(2);
    expect(allBooks).toEqual(expect.arrayContaining([book1, book2]));
  });
});
